### ISSUE TITLE
Android and iOS support

### ISSUE THREAD

**Issue by quininer** [ctz/rustls-native-certs#3]
I think we can look at how chromium is implemented.

https://github.com/chromium/chromium/blob/master/net/cert/cert_verify_proc_android.cc
https://github.com/chromium/chromium/blob/master/net/cert/cert_verify_proc_ios.cc

**Comment by djc** [comment]
Nope -- but if you're willing to contribute I'm happy to provide any guidance you need.

**Comment by buraktabn** [comment]
TBH, I was planning on it because of  https://github.com/paritytech/subxt/issues/563. However, using `webpki` seems to work on Android.

**Comment by cpu** [comment]
Android and iOS are both supported by https://github.com/rustls/rustls-platform-verifier

Is there user demand for a native-certs solution or should we suggest users consider using the platform verifier instead? That crate isn't quite ready for primetime but I also suspect we can get it ready sooner than we could add support for those platforms here.

**Comment by djc** [comment]
iOS support might be as easy as enabling the security-framework path for `ios` in addition to `macos`, I think?

But in general, the rustls-platform-verifier is probably a better alternative for most use cases.

**Comment by djc** [comment]
[rustls-platform-verifier](https://crates.io/crates/rustls-platform-verifier) has been released and is probably a better option in most cases.

**Comment by stormshield-gt** [comment]
`reqwest` already supports `rustls-native-certs` but not `rustls-platform-verifier` (https://github.com/seanmonstar/reqwest/pull/2286#discussion_r1609846678), blocked by https://github.com/rustls/rustls-platform-verifier/issues/58. 
https://github.com/rustls/rustls-platform-verifier/issues/58 seems harder to resolve that this current issue, Am I right ?

Do you think it will be reasonable to works on this issue first, enabling the usage of `reqwest` with `rustls` on android with loading the cert store ? I mean if you are open to contribution.
I know right now it's possible to embed a store with `rustls-webpki-roots` but it does not fit our need.

**Comment by djc** [comment]
Happy to review a PR for this but want to caution that if this needs too much complexity we might want to avoid it in favor of improving the platform verifier crate instead.

If you want to contribute, maybe you can contribute there?

**Comment by stormshield-gt** [comment]
Yes, hopefully I have some bandwidth to contribute here. Do you have any guidance on how things should be done?

**Comment by djc** [comment]
@stormshield-gt happy to discuss on Discord.

**Comment by djc** [comment]
As soon as we get to that level of complexity, it seems like we'd almost certainly be better off relying on the context/build infra/testing we already have in place in the rustls-platform-verifier repo.

**Comment by stormshield-gt** [comment]
Thanks a lot for the guidance. By checking things by my side, I reached to the same conclusion.  It seems like everybody agree that it would be better to look for supporting extra roots inside `rustls-plaform-verifier` to unlock the `reqwest` integration. I will put some effort in that direction.

**Comment by ctz** [comment]
On the android I just came across this in golang: https://github.com/golang/go/blob/master/src/crypto/x509/root_linux.go#L26-L30

**Comment by stormshield-gt** [comment]
For IOS, they used to embed the cert store statically and then switch to using the platform verifier https://github.com/golang/go/commit/b74f2efc47bbfcc4aa301ebda1033948d8b6b63e. That seems to confirm there is no API for loading the store on this platform

**Comment by cpu** [comment]
> On the android side I just came across this in golang: https://github.com/golang/go/blob/master/src/crypto/x509/root_linux.go#L26-L30

This seems like it could be reasonably adapted to this crate, but I echo @djc's concerns that when you start to consider test coverage for Android it would ratchet up the complexity of this repo's CI more than I think is deserved. 

Given there's no good solution for iOS I suspect most users invested in a "native" experience that includes mobile platforms will want to use `rustls-platform-verifier` anyway since only supporting Android is probably of limited utility. I'm inclined to say we should call this a WONTFIX for this repo and continue to try and invest in the platform verifier crate. Thoughts?

**Comment by djc** [comment]
(There was some follow-up in https://github.com/rustls/rustls-platform-verifier/issues/58#issuecomment-2299022565.)
